### PR TITLE
Improve error handling all around

### PR DESCRIPTION
**Pull request by whatyouhide**
* Added option validation for `EnvVar.Provider.init/1`
* Added verbose errors when types fail to parse

### CODE REVIEW

File: lib/providers/env_var_provider.ex [hunk 1]
```diff
@@ -58,27 +58,40 @@ defmodule EnvVar.Provider do
   in the config files for values that are configured in the EnvVar
   Provider.
   """
-  def init(prefix: prefix, env_map: env_map, enforce: enforce) when is_atom(prefix) do
```
Comment (whatyouhide) [inline lib/providers/env_var_provider.ex:61]: With this approach you're force to use this order when passing the options. With the new approach, you're not (which is the common way to do this in Elixir).

File: lib/providers/env_var_provider.ex [hunk 2]
```diff
@@ -156,7 +169,7 @@ defmodule EnvVar.Provider do
   defp get_env_value(key, config) do
     key
     |> System.get_env()
-    |> set_default(config[:default], key)
```
Comment (whatyouhide) [inline lib/providers/env_var_provider.ex:159]: This argument was not used.

File: lib/providers/env_var_provider.ex [hunk 3]
```diff
@@ -175,13 +188,17 @@ defmodule EnvVar.Provider do
...
+      {value, ""} -> value
+      _other -> raise ArgumentError, "expected float, got: #{inspect(env_value)}"
+    end
   end
 
   def convert(env_value, :integer) do
-    {val, _extra} = Integer.parse(env_value)
-    val
+    case Integer.parse(env_value) do
+      {value, ""} -> value
```
Comment (whatyouhide) [inline lib/providers/env_var_provider.ex:197]: We were allowing `123foo` which I think is not good. @relistan wdyt? Were we incidentally using it somewhere afayk?

File: lib/providers/env_var_provider.ex [hunk 4]
```diff
@@ -209,16 +226,12 @@ defmodule EnvVar.Provider do
     |> Enum.map(&convert(&1, type))
   end
 
-  defp set_default(value, default, _env_var_name) when is_nil(value) do
+  defp set_default(value, default) when is_nil(value) or value == "" do
     default
   end
 
-  defp set_default(value, default, _env_var_name) do
-    if String.length(value) < 1 do
```
Comment (whatyouhide) [inline lib/providers/env_var_provider.ex:217]: I think this is equivalent to `value == ""`, right?

**Review by uzarubin — APPROVED** [review]

**Comment by uzarubin** [comment]
Great changes all around!

File: lib/providers/env_var_provider.ex (hunk 1 above)
Comment (relistan) [reply lib/providers/env_var_provider.ex:61]: I'm definitely for this change. Wish there were a cleaner/nicer way to do this natively in Elixir :(

File: lib/providers/env_var_provider.ex [hunk 5]
```diff
@@ -58,27 +58,40 @@ defmodule EnvVar.Provider do
...
+      end
+
+    process_config(env_map, prefix, enforce?)
   end
 
-  def show_vars(prefix: prefix, env_map: env_map) when is_atom(prefix) do
+  def show_vars(opts) do
+    prefix = opts |> Keyword.fetch!(:prefix) |> String.to_atom()
+    env_map = Keyword.fetch!(opts, :env_map)
+
```
Comment (relistan) [inline lib/providers/env_var_provider.ex:87]: This doesn't enforce that both args are present and doesn't return any errors in that case. The previous implementation did, although the error was shitty.

File: lib/providers/env_var_provider.ex (hunk 2 above)
Comment (relistan) [reply lib/providers/env_var_provider.ex:159]: Leftover from previous implementation I think

File: lib/providers/env_var_provider.ex (hunk 3 above)
Comment (relistan) [reply lib/providers/env_var_provider.ex:197]: This is a good change. :shipit:

File: lib/providers/env_var_provider.ex [hunk 6]
```diff
@@ -209,16 +226,12 @@ defmodule EnvVar.Provider do
     |> Enum.map(&convert(&1, type))
   end
 
-  defp set_default(value, default, _env_var_name) when is_nil(value) do
+  defp set_default(value, default) when is_nil(value) or value == "" do
```
Comment (relistan) [inline lib/providers/env_var_provider.ex]: Elixir can tell the difference between an env var that is set, but the value is `""` and when it's not set at all (value of `nil`). We should NOT combine those behaviors.

e.g. 

```
$ EXISTS="" iex
...
iex(1)> System.get_env "EXISTS"
""
```
vs
```
iex(2)> System.get_env "DOES_NOT_EXIST"
nil
```

File: lib/providers/env_var_provider.ex (hunk 4 above)
Comment (relistan) [reply lib/providers/env_var_provider.ex:217]: Hmm, I'm confused about the previous behavior for a variable that is set but has empty string now.

**Review by relistan — APPROVED** [review]
One question around the behavior of empty string vs unset env vars. Otherwise looks great.

File: lib/providers/env_var_provider.ex (hunk 6 above)
Comment (relistan) [reply lib/providers/env_var_provider.ex]: Looks like we may previously have treated them the same way, but we shouldn't.

File: lib/providers/env_var_provider.ex (hunk 5 above)
Comment (whatyouhide) [reply lib/providers/env_var_provider.ex:87]: Wait I'm confused. This enforces that both the `:prefix` and `:env_map` options are there. Are referring to something else as well? 🤔

File: lib/providers/env_var_provider.ex (hunk 6 above)
Comment (whatyouhide) [reply lib/providers/env_var_provider.ex]: > Elixir can tell the difference between an env var that is set, but the value is "" and when it's not set at all (value of nil)

Oh I know, this change was just a simpler refactor of the previous behaviour. I agree we should treat them differently. Just to confirm, you want an empty string to be an empty string and not to fall back to the default value, correct?

Comment (relistan) [reply lib/providers/env_var_provider.ex]: Yeah, that seems like the right behavior.

File: lib/providers/env_var_provider.ex (hunk 5 above)
Comment (relistan) [reply lib/providers/env_var_provider.ex:87]: Sorry, forgot that `!` means fetch raises.

**Comment by whatyouhide** [comment]
@relistan updated, had to updated a bunch of tests too, would appreciate another quick look from you 💟

File: test/env_var_provider_test.exs [hunk 7]
```diff
@@ -64,14 +64,14 @@ defmodule EnvVar.ProviderTest do
...
     on_exit(fn ->
-      System.put_env("BEOWULF_MYCLUSTER_CLUSTER_OPTIONS_CREDENTIALS", "")
-      System.put_env("BEOWULF_MYCLUSTER_CLUSTER_OPTIONS_PORT", "")
-      System.put_env("BEOWULF_MYCLUSTER_CLUSTER_OPTIONS_LIST_KEY", "")
-      System.put_env("BEOWULF_MYCLUSTER_SERVER_COUNT", "")
-      System.put_env("BEOWULF_MYCLUSTER_NAME", "")
-      System.put_env("BEOWULF_MYCLUSTER_SETTINGS", "")
-      System.put_env("BEOWULF_MYCLUSTER_KEYS", "")
-      System.put_env("BEOWULF_APP_ENVVAR_PROVIDER", "")
+      System.delete_env("BEOWULF_MYCLUSTER_CLUSTER_OPTIONS_CREDENTIALS")
```
Comment (relistan) [inline test/env_var_provider_test.exs:67]: 👍

**Review by relistan — APPROVED** [review]
Lookin' good!

**Pull request merged**